resources/prosody-plugins/poltergeist.lib.lua

```diff
@@ -90,43 +90,61 @@ end
 local function get_username(room, user_id)
     local room_name = jid.node(room.jid)
 
-    if not usernames[room_name] then
+    if not state[room_name] then
         return nil
     end
 
-    return usernames[room_name][user_id]
+    return state[room_name][user_id]
 end
 
 local function get_username_from_nick(room_name, nick)
-    if not usernames[room_name] then
+    if not state[room_name] then
         return nil
     end
 
-    local user_id = usernames[room_name][nick]
-    return usernames[room_name][user_id]
+    local user_id = state[room_name][nick]
+    return state[room_name][user_id]
 end
 
 -- Removes the username from the store.
 -- @param room is the room the poltergeist is being removed from
 -- @param nick is the nick of the muc occupant
 local function remove_username(room, nick)
-    local room_name = jid.node(room.jid);
-    if not usernames[room_name] then
+    local room_name = jid.node(room.jid)
+    if not state[room_name] then
         return
     end
 
-    local user_id = usernames[room_name][nick]
-    usernames[room_name][user_id] = nil
-    usernames[room_name][nick] = nil
+    local user_id = state[room_name][nick]
+    state[room_name][user_id] = nil
+    state[room_name][nick] = nil
 end
 
 -- Removes all poltergeists in the store for the provided room.
 -- @param room is the room all poltergiest will be removed from
 local function remove_room(room)
     local room_name = jid.node(room.jid)
-    if usernames[room_name] then
-        usernames[room_name] = nil
+    if state[room_name] then
+        state[room_name] = nil
+    end
+end
+
+local function add_call_resource(room, call_id, resource_type, resource_id)
+    local room_name = jid.node(room.jid)
+    if not state[room_name] then
+        state[room_name] = {}
+        state[room_name][call_id] = {}
+        state[room_name][call_id][resource_type] = resource_id
+        return
+    end
+
+    if not state[room_name][call_id] then
+        state[room_name][call_id] = {}
+        state[room_name][call_id][resource_type] = resource_id
+        return
     end
+
+    state[room_name][call_id][resource_type] = resource_id
 end
```

Comment:
Can we simplify this a little to avoid duplication, like:

    if not state[room_name] then
        state[room_name] = {}
    end

    if not state[room_name][call_id] then
        state[room_name][call_id] = {}
    end

    state[room_name][call_id][resource_type] = resource_id

Reply:
Yes; done.